### PR TITLE
(#420) updated saving/added highlighting on partially mapped

### PR DESCRIPTION
**Pull request by shlag3n**
#420

1. updated mapping saving (will save what is possible and always on), mark mapping done is possible only for lead mapper + if no partial mappings
2. added highlighting to partially mapped items, **important:** those items will not be included to saving&mark mapping as done
3. improved error messaging
4. fixed blank screen on changelog/back buttons
5. fixed saving mappings for shared mappings
6. fixed some react related issues with mutating state

### CODE REVIEW

**Review by excelsior — CHANGES_REQUESTED** [review]
There are some issues (although not all of them belong to the scope of the task, I guess):

1. Saving a mapping with fully unmapped terms gives this error:

Screenshot 2024-02-21 at 10 23 34

2. If the above alert is closed, no subsequent errors are shown.
3. It doesn't seem possible to unmap a term.
4. Alignments with a `No Match` predicate are highlighted as partial (but get saved as expected).
5. It's possible to save a `No Match` alignment with a mapped term, though.
6. As far as I can see, partial alignments do get saved, so the success message is a bit misleading.

File: app/controllers/api/v1/alignments_controller.rb [hunk 1]
```diff
@@ -30,9 +30,15 @@ def create
                               .permit(alignments: [:id, :predicate_id, { mapped_term_ids: [] }])
                               .require(:alignments)
 
-        mapping = current_user.mappings.find(params.fetch(:mapping_id))
-        SaveAlignments.call(alignments: alignments_params, mapping: mapping)
-        head :ok
+        mapping = policy_scope(Mapping).find_by(id: params.fetch(:mapping_id))
+        raise Pundit::NotAuthorizedError, "not allowed to update this mapping" unless mapping.present?
+
+        interactor = SaveAlignments.call(alignments: alignments_params, mapping: mapping)
```
[the remark this answers is not in the document]
Comment (shlag3n) [reply app/controllers/api/v1/alignments_controller.rb:36]: good point, will address this separately and rebase this branch with main after that

**Comment by shlag3n** [comment]
> There are some issues (although not all of them belong to the scope of the task, I guess):
> 
> 1. Saving a mapping with fully unmapped terms gives this error:
> 2. If the above alert is closed, no subsequent errors are shown.
1, 2 - will fix and let you know

> 3. It doesn't seem possible to unmap a term.
We're collecting these things at notion - https://www.notion.so/learningtapestry/Q-A-1a06bee62f9248cb8751ff29efc21ce4, feel free to add there (this one already was added and it's not related to the task)

> 4. Alignments with a `No Match` predicate are highlighted as partial (but get saved as expected).
4 - will fix

> 5. It's possible to save a `No Match` alignment with a mapped term, though.
This is question at notion

> 6. As far as I can see, partial alignments do get saved, so the success message is a bit misleading.
Partial alignments will not be saved, only completed will be but you can click save at any moment and that was one of the requirement.

**Comment by shlag3n** [comment]
@excelsior p.2 was moved to https://github.com/t3-innovation-network/desm/issues/466

**Comment by excelsior** [comment]
Re: 6, partial alignments get saved but only under certain circumstances. Attached is a screencast demoing one of such cases (there are may be others).

https://github.com/t3-innovation-network/desm/assets/406292/b31e1105-6fe2-405f-b9d8-ca6b479a4efd

**Comment by excelsior** [comment]
I see. Everything else works as expected, thanks!

**Review by excelsior — APPROVED** [review]

**Pull request merged**